flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/FileStoreReadImpl.java

```diff
@@ -20,60 +20,132 @@
 
 import org.apache.flink.table.data.RowData;
 import org.apache.flink.table.data.binary.BinaryRowData;
+import org.apache.flink.table.runtime.generated.Projection;
 import org.apache.flink.table.store.file.FileFormat;
+import org.apache.flink.table.store.file.KeyValue;
 import org.apache.flink.table.store.file.mergetree.MergeTreeReader;
 import org.apache.flink.table.store.file.mergetree.compact.Accumulator;
 import org.apache.flink.table.store.file.mergetree.compact.IntervalPartition;
 import org.apache.flink.table.store.file.mergetree.sst.SstFileMeta;
 import org.apache.flink.table.store.file.mergetree.sst.SstFileReader;
 import org.apache.flink.table.store.file.utils.FileStorePathFactory;
 import org.apache.flink.table.store.file.utils.RecordReader;
+import org.apache.flink.table.store.utils.ProjectionUtils;
 import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.Comparator;
 import java.util.List;
 
 /** Default implementation of {@link FileStoreRead}. */
 public class FileStoreReadImpl implements FileStoreRead {
 
+    private final RowType keyType;
     private final SstFileReader.Factory sstFileReaderFactory;
     private final Comparator<RowData> keyComparator;
     private final Accumulator accumulator;
 
+    private Projection<RowData, BinaryRowData> keyProjection;
```

Comment:
store `int[][] projectedFields`, and create a separate `Projection` for  `createReader`.